[WPE] Implement save feature on HTTP RWI

magomez opened the pull request:
#### bd43e510a6385840dbf279f98374629fa3fc8d42
<pre>
[WPE] Implement save feature on HTTP RWI
<a href="https://bugs.webkit.org/show_bug.cgi?id=270636">https://bugs.webkit.org/show_bug.cgi?id=270636</a>

Reviewed by Devin Rousso.

Implement the save feature for single file in HTTP RWI.

* Source/WebInspectorUI/UserInterface/Base/BrowserInspectorFrontendHost.js:
(window.InspectorFrontendHost.WI.BrowserInspectorFrontendHost.prototype.canSave):
(window.InspectorFrontendHost.WI.BrowserInspectorFrontendHost.prototype.save):

Canonical link: <a href="https://commits.webkit.org/276097@main">https://commits.webkit.org/276097@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8caa53dd3d0ecd793935fe87ad706cc2bc5dce8c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43598 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22643 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46027 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46240 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39727 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26458 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20054 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/36032 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44172 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19666 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37545 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17001 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17221 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38623 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1655 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39767 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38917 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47788 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18636 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15232 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42818 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20056 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41488 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9729 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20236 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19687 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/BrowserInspectorFrontendHost.js:
```diff
@@ -226,13 +226,25 @@ if (!window.InspectorFrontendHost) {
+            console.assert(saveDatas.length === 1, saveDatas);
+
+            let saveData = saveDatas[0];
+            let blob = WI.BlobUtilities.blobForContent(saveData.content, saveData.base64Encoded, "octet/stream");
```
why `"octet/stream"`?

could/should we have that be a fallback for if something like `WI.mimeTypeForFileExtension(saveData.url)` doesn't work?

magomez replied:
"octet-stream" was a typo. Should have been "application/octet-stream". It's a fallback in case the download attribute of the <a> element doesn't work. It ensures that the browser will download the link instead of trying to open it, despite it may not be necessary in recent browsers.
I've changed it so we use WI.mimeTypeForFileExtension(saveData.url) as you suggested, and fallback to "application/octet-stream" if that function can't guess the mime type of the file (which happens for .har files).

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/BrowserInspectorFrontendHost.js:
```diff
@@ -226,13 +226,25 @@ if (!window.InspectorFrontendHost) {
+
+            let saveData = saveDatas[0];
+            let blob = WI.BlobUtilities.blobForContent(saveData.content, saveData.base64Encoded, "octet/stream");
+            let blobURL = window.URL.createObjectURL(blob);
```
NIT: you should be able to drop the `window` as we use `URL` directly elsewhere in Web Inspector

magomez replied:
Fixed.

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/BrowserInspectorFrontendHost.js:
```diff
@@ -226,13 +226,25 @@ if (!window.InspectorFrontendHost) {
+            a.style.display = "none";
+            // url looks like "web-inspector:///file.ext" use filename only
+            a.download = saveData.url.split("/").pop();
+            a.click();
```
will this work if `InspectorFrontendHost.save` was not called in response to user activation?

magomez replied:
I think it will work. I didn't test calling InspectorFrontendHost.save, but I've created a custom test that uses that very same approach to download a file from a setTimeout call, without user interaction, and it works as expected.

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/BrowserInspectorFrontendHost.js:
```diff
@@ -226,13 +226,25 @@ if (!window.InspectorFrontendHost) {
+            // url looks like "web-inspector:///file.ext" use filename only
+            a.download = saveData.url.split("/").pop();
+            a.click();
+            setTimeout(() => { window.URL.revokeObjectURL(blobURL); }, 0);
```
ditto about dropping `window`

you can also drop the `, 0` as that's implicit

Style: please put the body of non-implicit-return arrow functions on separate lines
```js
setTimeout(() => {
    URL.revokeObjectURL(blobURL);
});
```

magomez replied:
Fixed.

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/BrowserInspectorFrontendHost.js:
```diff
@@ -226,13 +226,25 @@ if (!window.InspectorFrontendHost) {
+            let a = document.createElement("a");
+            a.href = blobURL;
+            a.style.display = "none";
+            // url looks like "web-inspector:///file.ext" use filename only
```
Style: comments should be full sentences

magomez replied:
Fixed.

dcrousso reviewed (approved):
neat!  a few questions/concerns and style comments :)

magomez committed:
[WPE] Implement save feature on HTTP RWI
https://bugs.webkit.org/show_bug.cgi?id=270636

Reviewed by Devin Rousso.

Implement the save feature for single file in HTTP RWI.

* Source/WebInspectorUI/UserInterface/Base/BrowserInspectorFrontendHost.js:
(window.InspectorFrontendHost.WI.BrowserInspectorFrontendHost.prototype.canSave):
(window.InspectorFrontendHost.WI.BrowserInspectorFrontendHost.prototype.save):

Canonical link: https://commits.webkit.org/276097@main
